internal/character/clara/talent.go

```diff
@@ -0,0 +1,153 @@
+// equal to *% of Clara's ATK.
+//
+// flow:
+//
+// onBeforeBeingAttacked: + TalentMark, + TalentRes
+//
+// onAfterAttack: + follow-up attack
+func init() {
+	modifier.Register(TalentCounter, modifier.Config{
+		Listeners: modifier.Listeners{
+			// add marker modifier on attacker enemy
+			OnBeforeBeingAttacked: func(mod *modifier.Instance, e event.AttackStart) {
+				mod.Engine().AddModifier(e.Attacker, info.Modifier{
+					Name:   TalentMark,
+					Source: mod.Source(),
+				})
+			},
+
+			// remove marker modifier on all enemies when clara dies
+			OnBeforeDying: func(mod *modifier.Instance) {
+				for _, enemyID := range mod.Engine().Enemies() {
+					mod.Engine().RemoveModifier(enemyID, TalentMark)
+				}
+			},
+		},
+	})
+}
+
+func (c *char) initTalent() {
+	c.engine.AddModifier(c.id, info.Modifier{
+		Name:   TalentCounter,
+		Source: c.id,
+	})
+
+	c.engine.AddModifier(c.id, info.Modifier{
+		Name:   TalentRes,
+		Source: c.id,
+		Stats:  info.PropMap{prop.AllDamageReduce: 0.1},
+	})
+}
+
+// MAvatar_Klara_00_PassiveATK_Mark
+//
+// listens to AttackEnd events and check if clara is eligible for counter
+func (c *char) talentActionEndListener(e event.AttackEnd) {
+	attackerID := e.Attacker
+
+	// won't counter non-enemies (dominated ally)
+	if !c.engine.IsEnemy(attackerID) {
+		return
+	}
+
+	if c.canCounter(e) {
+		c.doCounter(attackerID)
+	}
+}
+
+// executes the follow-up attack,
+// NOTE: follow-up attack is same for both talent and e6
+func (c *char) doCounter(attackerID key.TargetID) {
+	c.engine.InsertAbility(info.Insert{
+		Execute: func() {
+			// DamageMaps
+			normalDamage := info.DamageMap{
+				model.DamageFormula_BY_ATK: talent[c.info.TalentLevelIndex()],
+			}
+			enhancedDamage := info.DamageMap{
+				model.DamageFormula_BY_ATK: talent[c.info.TalentLevelIndex()] +
+					ultDmgBoost[c.info.UltLevelIndex()],
+			}
+			mainTargetDamage := normalDamage
+
+			if c.engine.HasModifier(c.id, UltCounter) {
+				mainTargetDamage = enhancedDamage
+			}
+
+			// normal counter, damage
+			c.engine.Attack(info.Attack{
+				Source:       c.id,
+				Targets:      []key.TargetID{attackerID},
+				DamageType:   model.DamageType_PHYSICAL,
+				AttackType:   model.AttackType_INSERT,
+				BaseDamage:   mainTargetDamage,
+				StanceDamage: 30.0,
+				EnergyGain:   5.0,
+			})
+
+			// enhanced counter, damage to adjacent
+			// NOTE: in-game wording = damage value is based on main target's def
+			// mhy memes ?
+			if c.engine.HasModifier(c.id, UltCounter) {
+				splashDamage := info.DamageMap{model.DamageFormula_BY_ATK: (talent[c.info.TalentLevelIndex()] + ultDmgBoost[c.info.UltLevelIndex()]) * 0.5}
+
+				c.engine.Attack(info.Attack{
+					Source:       c.id,
+					Targets:      c.engine.AdjacentTo(attackerID),
+					DamageType:   model.DamageType_PHYSICAL,
+					AttackType:   model.AttackType_INSERT,
+					BaseDamage:   splashDamage,
+					StanceDamage: 30.0,
+				})
+			}
```

Comment:
Need to decrement the counter here and let the engine self-manage the modifier and delete it when count reaches 0